### PR TITLE
Add test to document request timeout behavior

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4462
```

### CODE REVIEW

**Comment by tkashem** [comment]
stress test result:
```
$ stress -p=32 ./server.test -test.run=TestRequestTimeoutBehavior
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 32 runs so far, 0 failures
20s: 32 runs so far, 0 failures
...
10m25s: 1701 runs so far, 0 failures
10m30s: 1720 runs so far, 0 failures
10m35s: 1733 runs so far, 0 failures
10m40s: 1751 runs so far, 0 failures
```

**Comment by tkashem** [comment]
sharing `stress` results (after rebasing onto master):
```
$ go test -race k8s.io/apiserver/pkg/server -c

$ stress -p=256 ./server.test -test.run=TestRequestTimeoutBehavior
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 1 runs so far, 0 failures
...
7m35s: 9963 runs so far, 0 failures
7m40s: 10085 runs so far, 0 failures
7m45s: 10197 runs so far, 0 failures
```

**Comment by tkashem** [comment]
this test adds about `10s` of overhead to the package `go test`
```
(master)
$ go test -count=1 -race k8s.io/apiserver/pkg/server
ok  	k8s.io/apiserver/pkg/server	78.053s

(this PR)
$ go test -count=1 -race k8s.io/apiserver/pkg/server
ok  	k8s.io/apiserver/pkg/server	88.275s
```

File: staging/src/k8s.io/apiserver/pkg/server/read_write_deadline_test.go [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+					clientStatusCodeExpected: 0,
+					clientRespBodyReadErr:    wantNoError{},
+					handlerWriteErr:          wantError{http.ErrHandlerTimeout},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		for _, proto := range []string{"HTTP/1.1", "HTTP/2.0"} { // every test is run for both http/1x and http/2.0
+			t.Run(fmt.Sprintf("%s/%s", test.setup.name, proto), func(t *testing.T) {
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/read_write_deadline_test.go:249]: these sub-tests can't be run in parallel using `t.Parallel'  when we run these tests with the feature gate `PerHandlerReadWriteTimeout` enabled/disabled

**Comment by tkashem** [comment]
> --- FAIL: TestWatchSemantics (1.62s)

looks like a knowm issue https://github.com/kubernetes/kubernetes/issues/125688

/test pull-kubernetes-unit

File: staging/src/k8s.io/apiserver/pkg/server/read_write_deadline_test.go (hunk 1 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/read_write_deadline_test.go:249]: I will implement the pattern for parallel subtests - ref: https://github.com/kubernetes/kubernetes/pull/122924

**Comment by sttts** [comment]
Looks a lot better now. Thanks!

/lgtm
/approve

**Pull request merged**